lsコマンドを作る4の提出

maruko92188 opened the pull request.

maruko92188 committed:
lsコマンドを作る4の提出

maruko92188 committed:
rubocopの指摘箇所を訂正

create_permissions(status)メソッドを分割

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,26 +2,126 @@
 def main
-  file_names = search_file_names
+  options = ARGV.getopts('alr')
+  search_file_names = options['a'] ? Dir['..', '.*', '*'] : Dir['*']
```
searchだと動詞なのでメソッドのように見えてしまいそうです
https://qiita.com/jnchito/items/459d58ba652bf4763820

maruko92188 replied:
searchedに変更しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,26 +2,126 @@
+
+def display_long_format(file_names)
+  long_formats = build_long_format_table(file_names)
+  total_blocks = long_formats.map { |format| format[:blocks] }.sum
```
```suggestion
  total_blocks = long_formats.sum { |format| format[:blocks] }
```

でもいけるはずです。

ただ、ここで変数宣言すると変数の寿命が無駄に長くなるかも？
https://qiita.com/jnchito/items/785e325debaca7185a2b

maruko92188 replied:
sumを使うように変更し、計算後すぐにtotalを表示させるようにしました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,26 +2,126 @@
+  total_blocks = long_formats.map { |format| format[:blocks] }.sum
+  widths = caluculate_max_length(long_formats)
+
+  rows = long_formats.map do |format|
```
ここでmapして57行目でeachすると2回ループが走ります。
ファイルが100万件あれば200万回ループが走ることになります。
ループ回数をできるだけ減らすアプローチを考えてみてください。

maruko92188 replied:
ループが1回で済むように変更しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,26 +2,126 @@
+      "#{format[:group_name].ljust(widths[:group_width])} ",
+      format[:byte_size].rjust(widths[:byte_width]).to_s,
+      format[:last_modified_time].to_s,
+      format[:path_name].to_s
```
to_sって必要なんでしょか？（joinするタイミングで自動的にto_sが呼ばれるはず）

maruko92188 replied:
rubocopの自動修正で直したところでした。
>文字列でない要素に対しては、to_str があれば to_str、なければ to_s した結果を連結します。

[Array\#join \(Ruby 3\.4 リファレンスマニュアル\)](https://docs.ruby-lang.org/ja/latest/method/Array/i/join.html)に、joinするときにto_sされるとありましたので削除しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,26 +2,126 @@
+end
+
+def build_long_format_table(file_names)
+  status_table = file_names.map { |file_name| [file_name, File.lstat(file_name)] }
```
ここも62行目のmapとあわせて2回ループが走りますね。

maruko92188 replied:
ループが1回で済むように変更しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,26 +2,126 @@
+end
+
+def apply_setuid(permissions)
+  permissions[2] = permissions[2].eql?('x') ? 's' : 'S'
```
引数の内容を破壊的に変更している点をできたら修正したいですね
https://bootcamp.fjord.jp/pages/do-not-use-destructive-method

maruko92188 replied:
特殊権限があるときは、新しく文字列を作りそれを呼び出すようにしました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,26 +2,126 @@
+end
+
+def create_last_modified_time(status)
+  half_a_year_seconds = (60 * 60 * 24) * (365.2425 / 2)
```
これは定数化してもいいかも？

maruko92188 replied:
定数化しました。

JunichiIto reviewed (commented):
コメントしました〜。

maruko92188 committed:
mainメソッド内の変数名を修正

search_file_namesをsearhed_file_namesに修正

maruko92188 committed:
mainメソッドの修正

修正前：sorted_file_names = search_file_names.sort_by(&:downcase)
修正後：sorted_file_names = searched_file_names.sort_by(&:downcase)

maruko92188 committed:
display_long_formatの修正

total_blocksの計算を、mapで配列を作らずにsumのみで行うように変更
繰り返し処理の回数を削減
不要なto_sの削除

maruko92188 committed:
build_long_format_tableの修正

繰り返し処理を削減

maruko92188 committed:
余分な空白を削除

maruko92188 committed:
create_permissionsの変更

破壊的に変更された引数が渡されることがないように修正

maruko92188 committed:
create_permissions内の変数名を変更

maruko92188 committed:
create_last_modified_timeの修正

半年の秒数計算を定数化
表示形式のみ判定を行うように変更

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,26 +2,126 @@
+end
+
+def display_long_format(file_names)
+  long_formats = build_long_format_table(file_names)
```
long_format_tableをbuildしたなら、

```suggestion
  long_format_table = build_long_format_table(file_names)
```

にした方が自然かなと思います
https://bootcamp.fjord.jp/pages/readable-variable-name

maruko92188 replied:
変数名を変更しました

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,26 +2,126 @@
+  total_blocks = long_formats.sum { |format| format[:blocks] }
+  puts "total #{total_blocks}"
+
+  widths = caluculate_max_length(long_formats)
```
lengthなのかwidthなのか、単数なのか複数なのか、きっちり揃えた方がいいかも
https://bootcamp.fjord.jp/pages/readable-variable-name

maruko92188 replied:
幅の計算なのでwidthに揃えました。
2つ以上の幅を計算しているので、メソッド名を複数形にしました。

maruko92188 replied:
このメソッド内で同じ処理が繰り返し書かれているのが気になったので、重複部分を別メソッドにしました。
1dd6b4f772b89e120d3287bc889930e6d2efbc1c

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,26 +2,126 @@
+  long_formats.each do |format|
+    rows = [
+      "#{format[:file_mode]} ",
+      format[:hard_links].rjust(widths[:link_width]),
```
```suggestion
      format[:hard_links].rjust(widths[:hard_links]),
```

widthは変数名から伝わるので、キー名を揃えてあげるとか

maruko92188 replied:
キー名を統一しました

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,26 +2,126 @@
+      hard_links: status.nlink.to_s,
+      owner_name: Etc.getpwuid(status.uid).name,
+      group_name: Etc.getgrgid(status.gid).name,
+      byte_size: status.blockdev? || status.chardev? ? format('%#x', status.rdev) : status.size.to_s,
```
```suggestion
      byte_size: byte_size,
```

事前に変数に入れておいた方が読みやすいかなーと思います。path_nameも同様です。
https://x.com/jnchito/status/1326468865103073280

maruko92188 replied:
変数に入れて判定するようにしました。
[サンプルコードでわかる！Ruby 3\.1の主な新機能と変更点 \#Ruby \- Qiita](https://qiita.com/jnchito/items/bcd9b7f59bf4b30ea5b3#%E3%83%8F%E3%83%83%E3%82%B7%E3%83%A5%E3%83%AA%E3%83%86%E3%83%A9%E3%83%AB%E3%81%A8%E3%82%AD%E3%83%BC%E3%83%AF%E3%83%BC%E3%83%89%E5%BC%95%E6%95%B0%E3%81%A7%E5%80%A4%E3%82%92%E7%9C%81%E7%95%A5%E3%81%A7%E3%81%8D%E3%82%8B%E3%82%88%E3%81%86%E3%81%AB%E3%81%AA%E3%81%A3%E3%81%9F)
こちらの記事を読んで、ローカル変数名とハッシュのキー名が同じなので値を省略しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,26 +2,126 @@
+    [status.sticky?, 't']
+  ]
+  octals.zip(special_bits_table).map do |octal, (is_special, character)|
+    standard_permissions = FILE_PERMISSIONS[octal.to_sym]
```
```suggestion
    standard_permissions = FILE_PERMISSIONS[octal]
```

シンボルに変換するぐらいなら、FILE_PERMISSIONSのキーを最初から文字列にしておくのもありかと。

maruko92188 replied:
to_sで文字列にしてからさらにto_symするのが冗長に感じたので、ハッシュのキーを文字列に変更しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,26 +2,126 @@
+  end.join
+end
+
+def apply_special_permission(standard_permissions, character)
```
characterだと「文字」なので、ちょっと抽象的過ぎますね。
もう少し具体性のある名前を付けてあげたいかも？
https://bootcamp.fjord.jp/articles/78

maruko92188 replied:
special_symbolに変更しました。

JunichiIto reviewed (commented):
コメントしました！

maruko92188 committed:
create_permissionsメソッド内の変数名を変更

characterをspecial_symbolに変更

maruko92188 committed:
display_long_formatメソッド内の変数名を変更

long_formatsをlong_format_tableに変更

maruko92188 committed:
caluculate_max_lengthメソッドの命名変更

caluculate_max_lengthからcaluculate_long_format_widthsに変更

maruko92188 committed:
タイプミスの修正

maruko92188 committed:
ハッシュのキー名を統一

maruko92188 committed:
build_long_format_tableメソッド内に中間変数を作成

判定が必要なbyte_sizeとpath_nameを一度変数に入れる

maruko92188 committed:
FILE_PERMISSIONSのキーを文字列に変更

maruko92188 committed:
calculate_long_format_widthsメソッドを修正

重複した処理をcalculate_max_widthメソッドとして定義
メソッド名calculate_long_format_widthsをcreate_long_format_widthsに変更

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,26 +2,132 @@
+  puts "total #{total_blocks}"
+
+  widths = create_long_format_widths(long_format_table)
+  long_format_table.each do |format|
+    rows = [
+      "#{format[:file_mode]} ",
```
formatって形式とか書式みたいな意味なので、`format[:file_mode]`だと「file_modeの書式」みたいな意味に見えちゃいそうです。
それよりもファイルの情報やファイルのステータス、と捉えた方が意味が伝わりやすいかも？


```suggestion
  file_status_table.each do |file_status|
    rows = [
      "#{file_status[:file_mode]} ",
```

JunichiIto reviewed (approved):
1点コメントしましたが、軽微な問題なのでこれでOKです！
ls5で修正しておいてください〜。
